### PR TITLE
Braininvader

### PR DESCRIPTION
**Pull request by kerventkm**

### CODE REVIEW

File: eeg_models/types.py [hunk 1]
```diff
@@ -5,7 +5,12 @@
...
+Any = Any
+Callable = Callable
+Device = torch.device
+Dict = Dict
 Directory = Path
 File = Path
+List = List
 npArray = NewType("nupmy array", np.ndarray)
-Device = torch.device
+Optional = Optional
```
Comment (v-goncharenko) [inline eeg_models/types.py]: remove this changes pls. They are not needed

File: eeg_models/eegnet.py [hunk 2]
```diff
@@ -72,7 +71,9 @@ def __init__(
             nn.AvgPool2d((1, rate_drops[1])),
             nn.Dropout(dropout_rate),
             nn.Flatten(),
-            nn.Linear(f2 * (n_samples // torch.prod(rate_drops)), n_classes),
+            nn.Linear(
+                f2 * (n_samples // (rate_drops[0] * rate_drops[1])), n_classes, False
+            ),
```
Comment (v-goncharenko) [inline eeg_models/eegnet.py]: remove this changes pls. They are not needed here

File: eeg_models/utils.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+def n_parameters(model: Any) -> int:
+    """Calculates number of parameters in the model.
+
+    Args:
+        model: Model to count parameters in.
+
+    Returns: Number of parameters in the model.
+    """
+    return sum(param.numel() for param in model.parameters())
```
Comment (v-goncharenko) [inline eeg_models/utils.py:41]: What is this function for?

You don't use it anywhere

**Review by v-goncharenko — CHANGES_REQUESTED** [review]

File: eeg_models/types.py (hunk 1 above)
Comment (kerventkm) [reply eeg_models/types.py]: done!

File: eeg_models/eegnet.py (hunk 2 above)
Comment (kerventkm) [reply eeg_models/eegnet.py]: done!

File: eeg_models/utils.py (hunk 3 above)
Comment (kerventkm) [reply eeg_models/utils.py:41]: It was for pytest (from previous branches) should I remove it too?

File: eeg_models/braininvaders.py [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+        root: Optional[Directory] = None,
+        split: str = "train",
+        transforms: Optional[Callable] = None,
+        transform: Optional[Callable] = None,
+        target_transform: Optional[Callable] = None,
+        download: bool = True,
+        NonAdaptive: bool = True,
+        Adaptive: bool = True,
+        Training: bool = True,
+        Online: bool = True,
```
Comment (v-goncharenko) [inline eeg_models/braininvaders.py]: rename to PEP 8 compatible names

File: eeg_models/braininvaders.py [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+                    path_folder, "subject{:d}".format(subject), "Session*", filename
+                )
+            )
+        return subject_paths
+
+    def __len__(self) -> int:
+        return len(self.raw_dataset)
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        return {"eegs": self.raw_dataset[index][0], "markers": self.raw_dataset[index][1]}
```
Comment (v-goncharenko) [inline eeg_models/braininvaders.py]: read and preprocess data at getitem function, no __init__

File: eeg_models/eegnet.py [hunk 6]
```diff
@@ -45,10 +45,10 @@ def __init__(
         kernel = rate // 2
         if f2 is None:
             f2 = d * f1
-        rate_drops = (4, 8)
+        rate_drops = torch.tensor([4, 8])
 
         super().__init__(
-            nn.Unflatten(0, (1, 1, n_channels)),
+            nn.Unflatten(1, (1, n_channels)),
```
Comment (v-goncharenko) [inline eeg_models/eegnet.py]: remove this changes

**Review by v-goncharenko — CHANGES_REQUESTED** [review]

File: eeg_models/eegnet.py (hunk 6 above)
Comment (kerventkm) [reply eeg_models/eegnet.py]: Done!

File: eeg_models/braininvaders.py (hunk 4 above)
Comment (kerventkm) [reply eeg_models/braininvaders.py]: Done!

File: eeg_models/braininvaders.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+                    path_folder, "subject{:d}".format(subject), "Session*", filename
+                )
+            )
+        return subject_paths
+
+    def __len__(self) -> int:
+        return len(self.raw_dataset)
+
+    def __getitem__(self, index: int) -> Dict[str, Any]:
+        self.m_data = self.get_data()
```
Comment (v-goncharenko) [inline eeg_models/braininvaders.py]: don't load all the data!

**Pull request closed without merging**